ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -74,10 +80,14 @@
   private final ConcurrentHashMap<String, AmbryDataNode> instanceNameToAmbryDataNode = new ConcurrentHashMap<>();
   private final ConcurrentHashMap<AmbryPartition, Set<AmbryReplica>> ambryPartitionToAmbryReplicas =
       new ConcurrentHashMap<>();
-  private final ConcurrentHashMap<AmbryDataNode, Set<AmbryReplica>> ambryDataNodeToAmbryReplicas =
+  private final ConcurrentHashMap<AmbryDataNode, ConcurrentHashMap<String, AmbryReplica>> ambryDataNodeToAmbryReplicas =
       new ConcurrentHashMap<>();
   private final ConcurrentHashMap<AmbryDataNode, Set<AmbryDisk>> ambryDataNodeToAmbryDisks = new ConcurrentHashMap<>();
   private final ConcurrentHashMap<ByteBuffer, AmbryPartition> partitionMap = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, ConcurrentHashMap<String, String>> partitionToResourceNameByDc =
+      new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, AtomicReference<RoutingTableSnapshot>> dcToRoutingTableSnapshotRef =
```

Comment:
Why do we need a AtomicReference here? Looks like we are setting this either in constructor or inside a lock.

Reply:
As discussed offline, we decided to keep the AtomicReference and remove the lock.

Reply:
@cgtz good point, I probably prefer to use `AtomicReference` and change it to regular `HashMap`.  Two reasons: (1) value in AtomicReference is volatile. (2) single CAS instruction, natively thread safe, fast and cheap, no context switch. What do you think?